src/settings/SettingsStore.js

```diff
@@ -0,0 +1,462 @@
+/*
+Copyright 2017 Travis Ralston
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import Promise from 'bluebird';
+import DeviceSettingsHandler from "./DeviceSettingsHandler";
+import RoomDeviceSettingsHandler from "./RoomDeviceSettingsHandler";
+import DefaultSettingsHandler from "./DefaultSettingsHandler";
+import RoomAccountSettingsHandler from "./RoomAccountSettingsHandler";
+import AccountSettingsHandler from "./AccountSettingsHandler";
+import RoomSettingsHandler from "./RoomSettingsHandler";
+import ConfigSettingsHandler from "./ConfigSettingsHandler";
+import {_t, _td} from '../languageHandler';
+import SdkConfig from "../SdkConfig";
+
+// Preset levels for room-based settings (eg: URL previews).
+// Doesn't include 'room' because most settings don't need it. Use .concat('room') to add.
+const LEVELS_PRESET_ROOM = ['device', 'room-device', 'room-account', 'account', 'config'];
+
+// Preset levels for account-based settings (eg: interface language).
+const LEVELS_PRESET_ACCOUNT = ['device', 'account', 'config'];
+
+// Preset levels for features (labs) settings.
+const LEVELS_PRESET_FEATURE = ['device', 'config'];
+
+const SETTINGS = {
+    // EXAMPLE SETTING:
+    // "my-setting": {
+    //     // Required by features, optional otherwise
+    //     isFeature: false,
+    //     displayName: _td("Cool Name"),
+    //
+    //     // Required.
+    //     supportedLevels: [
+    //         // The order does not matter.
+    //
+    //         "device",        // Affects the current device only
+    //         "room-device",   // Affects the current room on the current device
+    //         "room-account",  // Affects the current room for the current account
+    //         "account",       // Affects the current account
+    //         "room",          // Affects the current room (controlled by room admins)
+    //         "config",        // Affects the current application
+    //
+    //         // "default" is always supported and does not get listed here.
+    //     ],
+    //
+    //     // Optional. Any data type.
+    //     default: {
+    //         your: "value",
+    //     },
+    // },
+    "feature_groups": {
+        isFeature: true,
+        displayName: _td("Communities"),
+        supportedLevels: LEVELS_PRESET_FEATURE,
```

Comment:
As per #riot-dev (although i've lost the answer), I'm surprised that we're not specifying typing data about the settings: boolean, int, colour, string, etc... to aid drawing UI and reasoning about them in future

Reply:
> There's a wide variety of possible types, the theory is that the name would describe the type.

Was my answer, although it's not a very strong point. Should the "wrong" type being returned from a level cause that level to be skipped or an exception to be thrown? Should the code try and do any kind of massaging (`"11" -> 11`)?

Reply:
I wouldn't bother with massaging/validation, but it just seems wise to indicate the intention of the setting, especially for if/when we start autogenerating UI.  It may be premature though; up to you.